Datset links:
- https://www.kaggle.com/datasets/niharika41298/gym-exercise-data
- https://www.kaggle.com/datasets/sumedh1507/fitness-and-workout-dataset

In [ ]:
import pandas as pd
pd.set_option('display.max_colwidth', None)
pd.set_option('display.max_columns', 110)

In [54]:
program_df = pd.read_csv('fitness_and_workout_dataset.csv')
print(program_df.shape)

(2598, 10)


In [55]:
# drop created and last_edit columns that are not needed
program_df = program_df.drop(columns=['created', 'last_edit'])
program_df.head()

title  \
0  (MASS MONSTER) High Intensity 4 Day Upper Lower Program   
1                               (NOT MY PROGRAM)SHJ Jotaro   
2        1 PowerLift Per Day Powerbuilding 5 Day Bro Split   
3                            10 Week Mass Building Program   
4                                   10 week deadlift focus   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            description  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Build tones of muscular with this high intensity 4 day upper lower workout routine.\n\nWeekly Setup:\nDay 1 - Upper Push Focus\nDay 2 - Lower Quad Focus\nDay 3 - Rest/Active Recovery\nDay 4 - Upper Pull Focus\nDay 5 - Rest/Active Recovery\nDay 6 - Lower Hamstring Focus\nDay 7 - Rest/Active Recovery \n\nMuscle Groups |Sets Per Week\nBack - 10\nShoulders - 10\nChest - 9\nQuads - 9\nHamstrings - 9\nGlutes - 9\nBiceps - 7\nTriceps - 7\nLower Back - 4\nCalves - 4\nAbs - 4   
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         

In [56]:
program_df['equipment'].value_counts()

Full Gym         1848
Garage Gym        561
At Home           119
Dumbbell Only      69
Name: equipment, dtype: int64

In [57]:
exercise_df = pd.read_csv('megaGymDataset.csv')
print(exercise_df.shape)
exercise_df.head()

(2918, 9)


Unnamed: 0                         Title  \
0           0        Partner plank band row   
1           1  Banded crunch isometric hold   
2           2         FYR Banded Plank Jack   
3           3                 Banded crunch   
4           4                        Crunch   

                                                                                                                                                                                                                                                                                                                                                            Desc  \
0                                                                                                                                The partner plank band row is an abdominal exercise where two partners perform single-arm planks while pulling on the opposite ends of an exercise band. This technique can be done for time or reps in any ab-focused workout.   
1                                                                                                                                            The banded crunch isometric hold is an exercise targeting the abdominal muscles, particularly the rectus abdominis or "six-pack" muscles. The band adds resistance and continuous tension to this popular exercise.   
2                                     The banded plank jack is a variation on the plank that involves moving the legs in and out for repetitions. Having a band around the thighs forces the lower body to work harder, particularly the hips and glutes. The plank jack is commonly performed as part of a bodyweight circuit, or as part of a dynamic warm-up.   
3                                                                                                                                                           The banded crunch is an exercise targeting the abdominal muscles, particularly the rectus abdominis or "six-pack" muscles. The band adds resistance and continuous tension to this popular exercise.   
4  The crunch is a popular core exercise targeting the rectus abdominis, or "six-pack" muscles, as well as the obliques. It has been the centerpiece of many ab-focused workouts, due to both its simplicity and the intense burn and mind-muscle connection it produces. It can be performed for time or reps as part of the ab-focused portion of any workout.   

       Type    BodyPart Equipment         Level  Rating RatingDesc  
0  Strength  Abdominals     Bands  Intermediate     0.0        NaN  
1  Strength  Abdominals     Bands  Intermediate     NaN        NaN  
2  Strength  Abdominals     Bands  Intermediate     NaN        NaN  
3  Strength  Abdominals     Bands  Intermediate     NaN        NaN  
4  Strength  Abdominals     Bands  Intermediate     NaN        NaN

In [59]:
# Inspect body parts from exercise_df to know the categories
exercise_df['BodyPart'].value_counts()

Abdominals     662
Quadriceps     646
Shoulders      340
Chest          262
Biceps         168
Triceps        151
Lats           124
Hamstrings     121
Middle Back    118
Lower Back      97
Glutes          81
Calves          47
Forearms        31
Traps           24
Abductors       21
Adductors       17
Neck             8
Name: BodyPart, dtype: int64

In [60]:
import re

# Compact keyword -> BodyPart mapping
BODY_PART_KEYWORDS = {
    # Core
    **dict.fromkeys(['abs', 'core', 'abdominal', 'six pack', 'oblique'], 'Abdominals'),
    # Legs
    **dict.fromkeys(['quad', 'quads', 'leg', 'legs', 'thigh', 'squat', 'lunge', 'lower'], 'Quadriceps'),
    **dict.fromkeys(['hamstring', 'hamstrings', 'deadlift', 'posterior'], 'Hamstrings'),
    **dict.fromkeys(['glute', 'glutes', 'butt'], 'Glutes'),
    **dict.fromkeys(['calf', 'calves'], 'Calves'),
    # Upper body
    **dict.fromkeys(['chest', 'pec', 'pecs', 'bench', 'push up', 'push-up', 'push'], 'Chest'),
    **dict.fromkeys(['lats', 'lat', 'pull up', 'pull-up', 'pull'], 'Lats'),
    **dict.fromkeys(['upper back', 'middle back', 'mid back', 'back', 'row'], 'Middle Back'),
    **dict.fromkeys(['lower back'], 'Lower Back'),
    **dict.fromkeys(['shoulder', 'shoulders', 'delts', 'deltoid', 'press', 'overhead', 'upper'], 'Shoulders'),
    # Arms
    **dict.fromkeys(['bicep', 'biceps', 'curl'], 'Biceps'),
    **dict.fromkeys(['tricep', 'triceps', 'dip', 'skullcrusher'], 'Triceps'),
    **dict.fromkeys(['forearm', 'forearms'], 'Forearms'),
    # Other
    **dict.fromkeys(['trap', 'traps', 'shrug'], 'Traps'),
    **dict.fromkeys(['neck'], 'Neck'),
    **dict.fromkeys(['adductor', 'adductors'], 'Adductors'),
    **dict.fromkeys(['abductor', 'abductors'], 'Abductors'),
}

def infer_bodyparts_from_text(text: str):
    if pd.isna(text):
        return set()
    t = text.lower()
    found = set()
    for kw, bp in BODY_PART_KEYWORDS.items():
        # Use word boundaries to avoid partial matches
        pattern = r'\b' + re.escape(kw) + r'\b'
        if re.search(pattern, t):
            found.add(bp)
    return found

def infer_program_bodyparts(row):
    title = row.get('title', '')
    desc = row.get('description', '')
    parts = infer_bodyparts_from_text(str(title)) | infer_bodyparts_from_text(str(desc))
    return sorted(parts)

program_df['target_bodyparts'] = program_df.apply(infer_program_bodyparts, axis=1)
program_df[['title', 'description','target_bodyparts']].sample(50)

title  \
30        2 Day Ultra Busy Split: Arms, Legs, Core Focus   
2050                                     Size N Strength   
510                                        Bigga Workout   
1891                                           Rabid Dog   
1272                                      Juggernaut 2.0   
1366                                         Lilys planr   
1914                                       Redemption V2   
1181                                          INVINCIBLE   
1337                                        Lean & Built   
721                                              Dad Bod   
1091                                 Hezbollah Destroyer   
619                           Bullmastiff Custom Phase 3   
1897                          Ramping Basics Upper Lower   
772   Dumbbells for Life: A Strength-Based Workout Cycle   
656                      Chat GPT PPL - Upper body Focus   
2585                         “ATTITUDE ERA" WWE PHYSIQUE   
1439                               Matlock’s upper lower   
2289                                      Tom's Strenght   
2523                                   awsom 4 day split   
956                                        GZCL adjusted   
1159                        Hypertrophy : 4 Weeks 4 Days   
758                                Doug Hepburn Original   
1566                                   My simple program   
371          BIKINI BODY Muscle Growth Training Plan (1)   
1084                                            Hercules   
1558                                My Upper/Lower Split   
56               3 Days 20Lbs in 12 Week Workout Routine   
1776                                    Powerbuilding V1   
1609   Nuckols Stronger by Science Deadlift 3X High Vol.   
1733                                      Phoenix Reborn   
1039                         H12: Hypertrophy & Strength   
2234         The Fitpocalypse full-body training program   
843                                       Feather mark 1   
859                                        First Program   
1379       Low Volume High Intensity Program for Natty's   
1704                                 PROJECT SNATCHED !!   
2362                              Ultimate Upper Lower +   
546                            Body Weight Home Training   
1873                                      REBUILD SYSTEM   
2342                                     UL PPL Strength   
1699                                       PPSA 70’s Big   
2229                                The Dumbbell Odyssey   
1631                                        Oliver final   
2469                                          Winter Ark   
1055                                   HSP Periodization   
539                                                 Body   
1203                         Iron Monk Powerbuilding U/L   
622                                     Busyman Strength   
2124        Strength and hypertrophy basic powerbuilding   
1216           JKs Program - Vol 1 (5/3/1 Powerbuilding)   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              

#### After mapping, ends up 2598-1559 = 1039 workout program with specific body parts mapped

In [61]:
program_df['target_bodyparts'].value_counts()

[]                                                                       1559
[Quadriceps, Shoulders]                                                   177
[Quadriceps]                                                               65
[Chest]                                                                    50
[Chest, Lats, Quadriceps]                                                  44
                                                                         ... 
[Glutes, Middle Back, Quadriceps]                                           1
[Biceps, Shoulders, Triceps]                                                1
[Biceps, Chest, Hamstrings, Middle Back, Neck, Quadriceps, Shoulders]       1
[Abdominals, Chest, Middle Back, Quadriceps]                                1
[Chest, Neck, Traps]                                                        1
Name: target_bodyparts, Length: 230, dtype: int64

In [64]:
program_df_1 = program_df[program_df['target_bodyparts'].map(len) > 0]
print(program_df_1.shape)

(1039, 9)


In [68]:
program_df_1['target_bodyparts'].value_counts()

[Quadriceps, Shoulders]                                                  177
[Quadriceps]                                                              65
[Chest]                                                                   50
[Chest, Lats, Quadriceps]                                                 44
[Middle Back]                                                             36
                                                                        ... 
[Glutes, Middle Back, Quadriceps]                                          1
[Biceps, Shoulders, Triceps]                                               1
[Biceps, Chest, Hamstrings, Middle Back, Neck, Quadriceps, Shoulders]      1
[Abdominals, Chest, Middle Back, Quadriceps]                               1
[Chest, Neck, Traps]                                                       1
Name: target_bodyparts, Length: 229, dtype: int64

In [69]:
program_df_1.head()

title  \
0  (MASS MONSTER) High Intensity 4 Day Upper Lower Program   
2        1 PowerLift Per Day Powerbuilding 5 Day Bro Split   
4                                   10 week deadlift focus   
5                                  100 Push Ups in 6 Weeks   
7                                     100x Push Up’s / Day   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            description  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Build tones of muscular with this high intensity 4 day upper lower workout routine.\n\nWeekly Setup:\nDay 1 - Upper Push Focus\nDay 2 - Lower Quad Focus\nDay 3 - Rest/Active Recovery\nDay 4 - Upper Pull Focus\nDay 5 - Rest/Active Recovery\nDay 6 - Lower Hamstring Focus\nDay 7 - Rest/Active Recovery \n\nMuscle Groups |Sets Per Week\nBack - 10\nShoulders - 10\nChest - 9\nQuads - 9\nHamstrings - 9\nGlutes - 9\nBiceps - 7\nTriceps - 7\nLower Back - 4\nCalves - 4\nAbs - 4   
2  Based off of Andy Baker's KCS (Kingwood Strength & Conditioning) Powerbuilding program.\n\nIt focuses on increasing strength on one lift per day, including squat, deadlift, bench press, overhead press, and pull-up. Rest days are after squat and deadlift days.\n\n1 - Shoulders & Triceps\n2 - Back\n3 - Quads & Calves\nRest\n4 - Chest & Biceps\n5 - Posterior Chain\nRest\n\nFit in Abs, extra calves & Cardio wherever you like.\n\nAt some point in a lifter's career, typically after the novice or early intermediate phase, the most reliable way to get stronger is to become more muscular, i.e. build muscle mass. To that end, the term "powerbuilding" has emerged to describe a hybrid approach of training the main barbell lifts (like a powerlifter) alongside a selection of supplemental lifts done for higher reps aimed at increasing muscular size (like a bodybuilder). \n\nWhile some people knock powerbuilding as merely a clever market

### Phase 2 mapping - based on level, equipment, and bodyparts, to map exercises to workout program

In [ ]:
import ast

def parse_list_column(series):
    parsed = []
    for val in series.fillna('[]'):
        try:
            if isinstance(val, list):
                parsed.append(val)
            else:
                parsed.append(ast.literal_eval(val))
        except (ValueError, SyntaxError):
            parsed.append([val] if val not in [None, 'nan', ''] else [])
    return parsed

# Parse level column
if 'level_list' not in program_df_1.columns:
    program_df_1['level_list'] = parse_list_column(program_df_1['level'])

program_df_1[['title', 'level_list', 'equipment', 'target_bodyparts']].head()

/var/folders/wx/2vm3xplj00z38ycg7_mg727c0000gn/T/ipykernel_47880/1365350053.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  program_df_1['level_list'] = parse_list_column(program_df_1['level'])


title  \
0  (MASS MONSTER) High Intensity 4 Day Upper Lower Program   
2        1 PowerLift Per Day Powerbuilding 5 Day Bro Split   
4                                   10 week deadlift focus   
5                                  100 Push Ups in 6 Weeks   
7                                     100x Push Up’s / Day   

                         level_list equipment  \
0                    [Intermediate]  Full Gym   
2  [Beginner, Novice, Intermediate]  Full Gym   
4          [Intermediate, Advanced]  Full Gym   
5            [Novice, Intermediate]   At Home   
7            [Novice, Intermediate]   At Home   

                                                                                                         target_bodyparts  
0  [Abdominals, Biceps, Calves, Chest, Glutes, Hamstrings, Lats, Lower Back, Middle Back, Quadriceps, Shoulders, Triceps]  
2                      [Abdominals, Biceps, Calves, Chest, Hamstrings, Lats, Middle Back, Quadriceps, Shoulders, Triceps]  
4                                                                                                            [Hamstrings]  
5                                                                                                                 [Chest]  
7                                                                                                                 [Chest]

In [ ]:
# Level compatibility: map exercise Level to program level_list
def exercise_level_to_prog_labels(ex_level: str):
    """Map exercise Level (Beginner/Intermediate/Expert) to program level labels."""
    if pd.isna(ex_level):
        return []
    ex_level = str(ex_level).strip()
    if ex_level == 'Beginner':
        return ['Beginner', 'Novice']
    elif ex_level == 'Intermediate':
        return ['Intermediate']
    elif ex_level == 'Expert':
        return ['Advanced']
    return []

def level_compatible(program_levels, exercise_level):
    """Check if exercise level is compatible with program levels."""
    pl = set(program_levels or [])
    if not pl:  # if no program levels specified, accept all
        return True
    needed = set(exercise_level_to_prog_labels(exercise_level))
    if not needed:  # if exercise level unknown, accept
        return True
    return len(pl & needed) > 0

# Equipment compatibility
def equipment_compatible(program_equipment, exercise_equipment):
    """Check if exercise equipment is compatible with program equipment."""
    if pd.isna(program_equipment):
        return True
    
    pe = str(program_equipment).strip()
    ee = str(exercise_equipment).strip()
    
    # Full Gym and Garage Gym accept all equipment
    if pe in ['Full Gym', 'Garage Gym']:
        return True
    
    # Dumbbell Only accepts: Dumbbell, Kettlebells, Body Only
    if pe == 'Dumbbell Only':
        return ee in ['Dumbbell', 'Kettlebells', 'Body Only', 'None']
    
    # At Home accepts: Body Only, Bands, Exercise Ball, Foam Roll, Dumbbell
    if pe == 'At Home':
        return ee in ['Body Only', 'Bands', 'None', 'Exercise Ball', 'Foam Roll', 'Dumbbell']
    
    return True

Compatibility functions defined.


In [91]:
# Build program -> exercises mapping
program_exercise_mapping = []

for idx, prow in program_df_1.iterrows():
    p_id = idx
    p_title = prow['title']
    p_levels = prow.get('level_list', []) or []
    p_equip = prow['equipment']
    p_bodyparts = set(prow.get('target_bodyparts', []))
    
    # Filter exercises by body parts
    candidates = exercise_df[exercise_df['BodyPart'].isin(p_bodyparts)].copy()
    
    # Filter by level compatibility
    candidates = candidates[
        candidates['Level'].apply(lambda lv: level_compatible(p_levels, lv))
    ]
    
    # Filter by equipment compatibility
    candidates = candidates[
        candidates['Equipment'].apply(lambda eq: equipment_compatible(p_equip, eq))
    ]
    
    # Collect exercise info
    for _, ex_row in candidates.iterrows():
        program_exercise_mapping.append({
            'program_id': p_id,
            'program_title': p_title,
            'program_description': prow['description'],
            'program_goal': prow.get('goal', ''),
            'program_level': ', '.join(p_levels),
            'program_equipment': p_equip,
            'program_bodyparts': ', '.join(sorted(p_bodyparts)),
            'exercise_id': int(ex_row['Unnamed: 0']),
            'exercise_title': ex_row['Title'],
            'exercise_bodypart': ex_row['BodyPart'],
            'exercise_equipment': ex_row['Equipment'],
            'exercise_level': ex_row['Level'],
            'exercise_rating': ex_row['Rating'] if pd.notna(ex_row['Rating']) else 0.0
        })
    
    if idx % 100 == 0:
        print(f"Processed {idx} programs...")

program_exercise_df = pd.DataFrame(program_exercise_mapping)
print(f"\nCreated {len(program_exercise_df):,} program-exercise mappings")
program_exercise_df.head(5)

Processed 0 programs...
Processed 100 programs...
Processed 500 programs...
Processed 800 programs...
Processed 1100 programs...
Processed 1200 programs...
Processed 1400 programs...
Processed 1500 programs...
Processed 1600 programs...
Processed 1700 programs...
Processed 2400 programs...

Created 708,953 program-exercise mappings


program_id                                            program_title  \
0           0  (MASS MONSTER) High Intensity 4 Day Upper Lower Program   
1           0  (MASS MONSTER) High Intensity 4 Day Upper Lower Program   
2           0  (MASS MONSTER) High Intensity 4 Day Upper Lower Program   
3           0  (MASS MONSTER) High Intensity 4 Day Upper Lower Program   
4           0  (MASS MONSTER) High Intensity 4 Day Upper Lower Program   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                       program_description  \
0  Build tones of muscular with this high intensity 4 day upper lower workout routine.\n\nWeekly Setup:\nDay 1 - Upper Push Focus\nDay 2 - Lower Quad Focus\nDay 3 - Rest/Active Recovery\nDay 4 - Upper Pull Focus\nDay 5 - Rest/Active Recovery\nDay 6 - Lower Hamstring Focus\nDay 7 - Rest/Active Recovery \n\nMuscle Groups |Sets Per Week\nBack - 10\nShoulders - 10\nChest - 9\nQuads - 9\nHamstrings - 9\nGlutes - 9\nBiceps - 7\nTriceps - 7\nLower Back - 4\nCalves - 4\nAbs - 4   
1  Build tones of muscular with this high intensity 4 day upper lower workout routine.\n\nWeekly Setup:\nDay 1 - Upper Push Focus\nDay 2 - Lower Quad Focus\nDay 3 - Rest/Active Recovery\nDay 4 - Upper Pull Focus\nDay 5 - Rest/Active Recovery\nDay 6 - Lower Hamstring Focus\nDay 7 - Rest/Active Recovery \n\nMuscle Groups |Sets Per Week\nBack - 10\nShoulders - 10\nChest - 9\nQuads - 9\nHamstrings - 9\nGlutes - 9\nBiceps - 7\nTriceps - 7\nLower Back - 4\nCalves - 4\nAbs - 4   
2  Build tones of muscular with this high intensity 4 day upper lower workout routine.\n\nWeekly Setup:\nDay 1 - Upper Push Focus\nDay 2 - Lower Quad Focus\nDay 3 - Rest/Active Recovery\nDay 4 - Upper Pull Focus\nDay 5 - Rest/Active Recovery\nDay 6 - Lower Hamstring Focus\nDay 7 - Rest/Active Recovery \n\nMuscle Groups |Sets Per Week\nBack - 10\nShoulders - 10\nChest - 9\nQuads - 9\nHamstrings - 9\nGlutes - 9\nBiceps - 7\nTriceps - 7\nLower Back - 4\nCalves - 4\nAbs - 4   
3  Build tones of muscular with this high intensity 4 day upper lower workout routine.\n\nWeekly Setup:\nDay 1 - Upper Push Focus\nDay 2 - Lower Quad Focus\nDay 3 - Rest/Active Recovery\nDay 4 - Upper Pull Focus\nDay 5 - Rest/Active Recovery\nDay 6 - Lower Hamstring Focus\nDay 7 - Rest/Active Recovery \n\nMuscle Groups |Sets Per Week\nBack - 10\nShoulders - 10\nChest - 9\nQuads - 9\nHamstrings - 9\nGlutes - 9\nBiceps - 7\nTriceps - 7\nLower Back - 4\nCalves - 4\nAbs - 4   
4  Build tones of muscular with this high intensity 4 day upper lower workout routine.\n\nWeekly Setup:\nDay 1 - Upper Push Focus\nDay 2 - Lower Quad Focus\nDay 3 - Rest/Active Recovery\nDay 4 - Upper Pull Focus\nDay 5 - Rest/Active Recovery\nDay 6 - Lower Hamstring Focus\nDay 7 - Rest/Active Recovery \n\nMuscle Groups |Sets Per Week\nBack - 10\nShoulders - 10\nChest - 9\nQuads - 9\nHamstrings - 9\nGlutes - 9\nBiceps - 7\nTriceps - 7\nLower Back - 4\nCalves - 4\nAbs - 4   

                                   program_goal program_level  \
0  ['Muscle & Sculpting', 'Bodyweight Fitness']  Intermediate   
1  ['Muscle & Sculpting', 'Bodyweight Fitness']  Intermediate   
2  ['Muscle & Sculpting', 'Bodyweight Fitness']  Intermediate   
3  ['Muscle & Sculpting', 'Bodyweight Fitness']  Intermediate   
4  ['Muscle & Sculpting', 'Bodyweight Fitness']  Intermediate   

  program_equipment  \
0          Full Gym   
1          Full Gym   
2          Full Gym   
3          Full Gym   
4          Full Gym   

                                                                                                      program_bodyparts  \
0  Abdominals, Biceps, Calves, Chest

In [93]:
program_exercise_df['exercise_rating'].value_counts()

0.0    544146
8.8     12213
8.3     10160
9.0      9670
8.5      9629
        ...  
6.6        68
3.5        57
3.1        45
2.8        23
2.5        19
Name: exercise_rating, Length: 71, dtype: int64

#### Abandon all records where its exercise doesn't have a rating

In [94]:
# filter out all mappings with exercise_rating of 0.0
final_df = program_exercise_df[program_exercise_df['exercise_rating'] > 0.0]
final_df.head()

program_id                                            program_title  \
8            0  (MASS MONSTER) High Intensity 4 Day Upper Lower Program   
9            0  (MASS MONSTER) High Intensity 4 Day Upper Lower Program   
10           0  (MASS MONSTER) High Intensity 4 Day Upper Lower Program   
11           0  (MASS MONSTER) High Intensity 4 Day Upper Lower Program   
18           0  (MASS MONSTER) High Intensity 4 Day Upper Lower Program   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                        program_description  \
8   Build tones of muscular with this high intensity 4 day upper lower workout routine.\n\nWeekly Setup:\nDay 1 - Upper Push Focus\nDay 2 - Lower Quad Focus\nDay 3 - Rest/Active Recovery\nDay 4 - Upper Pull Focus\nDay 5 - Rest/Active Recovery\nDay 6 - Lower Hamstring Focus\nDay 7 - Rest/Active Recovery \n\nMuscle Groups |Sets Per Week\nBack - 10\nShoulders - 10\nChest - 9\nQuads - 9\nHamstrings - 9\nGlutes - 9\nBiceps - 7\nTriceps - 7\nLower Back - 4\nCalves - 4\nAbs - 4   
9   Build tones of muscular with this high intensity 4 day upper lower workout routine.\n\nWeekly Setup:\nDay 1 - Upper Push Focus\nDay 2 - Lower Quad Focus\nDay 3 - Rest/Active Recovery\nDay 4 - Upper Pull Focus\nDay 5 - Rest/Active Recovery\nDay 6 - Lower Hamstring Focus\nDay 7 - Rest/Active Recovery \n\nMuscle Groups |Sets Per Week\nBack - 10\nShoulders - 10\nChest - 9\nQuads - 9\nHamstrings - 9\nGlutes - 9\nBiceps - 7\nTriceps - 7\nLower Back - 4\nCalves - 4\nAbs - 4   
10  Build tones of muscular with this high intensity 4 day upper lower workout routine.\n\nWeekly Setup:\nDay 1 - Upper Push Focus\nDay 2 - Lower Quad Focus\nDay 3 - Rest/Active Recovery\nDay 4 - Upper Pull Focus\nDay 5 - Rest/Active Recovery\nDay 6 - Lower Hamstring Focus\nDay 7 - Rest/Active Recovery \n\nMuscle Groups |Sets Per Week\nBack - 10\nShoulders - 10\nChest - 9\nQuads - 9\nHamstrings - 9\nGlutes - 9\nBiceps - 7\nTriceps - 7\nLower Back - 4\nCalves - 4\nAbs - 4   
11  Build tones of muscular with this high intensity 4 day upper lower workout routine.\n\nWeekly Setup:\nDay 1 - Upper Push Focus\nDay 2 - Lower Quad Focus\nDay 3 - Rest/Active Recovery\nDay 4 - Upper Pull Focus\nDay 5 - Rest/Active Recovery\nDay 6 - Lower Hamstring Focus\nDay 7 - Rest/Active Recovery \n\nMuscle Groups |Sets Per Week\nBack - 10\nShoulders - 10\nChest - 9\nQuads - 9\nHamstrings - 9\nGlutes - 9\nBiceps - 7\nTriceps - 7\nLower Back - 4\nCalves - 4\nAbs - 4   
18  Build tones of muscular with this high intensity 4 day upper lower workout routine.\n\nWeekly Setup:\nDay 1 - Upper Push Focus\nDay 2 - Lower Quad Focus\nDay 3 - Rest/Active Recovery\nDay 4 - Upper Pull Focus\nDay 5 - Rest/Active Recovery\nDay 6 - Lower Hamstring Focus\nDay 7 - Rest/Active Recovery \n\nMuscle Groups |Sets Per Week\nBack - 10\nShoulders - 10\nChest - 9\nQuads - 9\nHamstrings - 9\nGlutes - 9\nBiceps - 7\nTriceps - 7\nLower Back - 4\nCalves - 4\nAbs - 4   

                                    program_goal program_level  \
8   ['Muscle & Sculpting', 'Bodyweight Fitness']  Intermediate   
9   ['Muscle & Sculpting', 'Bodyweight Fitness']  Intermediate   
10  ['Muscle & Sculpting', 'Bodyweight Fitness']  Intermediate   
11  ['Muscle & Sculpting', 'Bodyweight Fitness']  Intermediate   
18  ['Muscle & Sculpting', 'Bodyweight Fitness']  Intermediate   

   program_equipment  \
8           Full Gym   
9           Full Gym   
10          Full Gym   
11          Full Gym   
18          Full Gym   

                                                                                                       program_bodyparts  \
8   Abdomina